common/blobstore/s3store/client.go

```diff
@@ -0,0 +1,321 @@
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/aws/aws-sdk-go/service/s3/s3iface"
+	"github.com/uber/cadence/common/backoff"
+	"github.com/uber/cadence/common/blobstore"
+	"github.com/uber/cadence/common/blobstore/blob"
+	"go.uber.org/multierr"
+)
+
+const (
+	defaultBlobstoreTimeout = 10 * time.Second
+)
+
+type client struct {
+	s3cli s3iface.S3API
+}
+
+// NewClient returns a new Client backed by S3
+func NewClient(s3cli s3iface.S3API) blobstore.Client {
+	return &client{
+		s3cli: s3cli,
+	}
+}
+func ClientFromConfig(cfg *Config) (s3iface.S3API, error) {
+	if err := cfg.Validate(); err != nil {
+		return nil, err
+	}
+
+	s3Config := &aws.Config{
+		Endpoint:         cfg.Endpoint,
+		Region:           aws.String(cfg.Region),
+		S3ForcePathStyle: aws.Bool(cfg.S3ForcePathStyle),
+		MaxRetries:       aws.Int(0),
+	}
+	sess, err := session.NewSession(s3Config)
+	if err != nil {
+		return nil, err
+	}
+	return s3.New(sess), nil
+}
+
+func (c *client) Upload(ctx context.Context, bucket string, key blob.Key, blob *blob.Blob) error {
+	ctx, cancel := c.ensureContextTimeout(ctx)
+	defer cancel()
+	params := url.Values{}
+	for k, v := range blob.Tags {
+		params.Add(k, v)
+	}
+
+	_, err := c.s3cli.PutObjectWithContext(ctx, &s3.PutObjectInput{
+		Bucket:  aws.String(bucket),
+		Key:     aws.String(key.String()),
+		Body:    bytes.NewReader(blob.Body),
+		Tagging: aws.String(params.Encode()),
+	})
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok {
+			if aerr.Code() == s3.ErrCodeNoSuchBucket {
+				return blobstore.ErrBucketNotExists
+			}
+		}
+		return err
+	}
+	return nil
+}
+
+func (c *client) Download(ctx context.Context, bucket string, key blob.Key) (*blob.Blob, error) {
+	ctx, cancel := c.ensureContextTimeout(ctx)
+	defer cancel()
+	result, err := c.s3cli.GetObjectWithContext(ctx, &s3.GetObjectInput{
+		Bucket: aws.String(bucket),
+		Key:    aws.String(key.String()),
+	})
+
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok {
+			if aerr.Code() == s3.ErrCodeNoSuchBucket {
+				return nil, blobstore.ErrBucketNotExists
+			}
+
+			if aerr.Code() == s3.ErrCodeNoSuchKey {
+				return nil, blobstore.ErrBlobNotExists
+			}
+		}
+		return nil, err
+	}
+
+	defer func() {
+		if ierr := result.Body.Close(); ierr != nil {
+			err = multierr.Append(err, ierr)
+		}
+	}()
+
+	body, err := ioutil.ReadAll(result.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	tags, err := getTags(ctx, c.s3cli, bucket, key.String())
+	if err != nil {
+		return nil, err
+	}
+	return blob.NewBlob(body, tags), nil
+}
+
+func (c *client) GetTags(ctx context.Context, bucket string, key blob.Key) (map[string]string, error) {
+	ctx, cancel := c.ensureContextTimeout(ctx)
+	defer cancel()
+	tags, err := getTags(ctx, c.s3cli, bucket, key.String())
+	if err != nil {
+		return nil, err
+	}
+	return tags, nil
+}
+
+// Exists returns false when the bucket does not exist or the bucket and key does not exist
```

Comment:
Filestore currently returns error if bucket does not exist (e.g. bucket not existing is an error but key not existing is just false, nil) I think both APIs are reasonable but I want them to be the same. Lets change this to return an error if bucket does not exist.

Reply:
Problem is it would be more inefficient. I.e. I would need to call a different api method. The same error is returned from the api whether or not the bucket exists. What do you think?